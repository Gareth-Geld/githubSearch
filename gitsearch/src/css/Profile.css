.profileContainer {
    width: 1200px;
    height: fit-content;
    margin-left: auto;
    margin-right: auto;
    background-color: #2c066f;
    box-shadow: 0 0 5px 5px #00ff43;
    border-radius: 10%;
}

.loadingContainer {
    background-color: #31323c;
    border-radius: 10%;
}

.input-group {
    visibility: show;
}

.avatarProfile {
    position: relative;
    top: 50px;
    width: 300px;
    height: 300px;
    border-radius: 50%;
}

.title {
    position: relative;
    font-size: 90px;
    font-family: Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif;
    text-align: center;
    top: 50%;
    left: -15%;
}

.repoTitle {
    font-size: 40px;
    font-family: Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif;
    margin-top: 50px;
    margin-bottom: 40px;
    text-align: left;
    margin-left: 50px;
}

.repoListTitle {
    font-size: 40px;
    font-family: Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif;
    margin-top: 50px;
    margin-bottom: 40px;
    text-align: center;
    margin-left: 50px;
}

.BioText {
    font-size: 20px;
    padding: 0px 20px 0px 20px
}

.userInfo {
    margin-top: 80px;
    font-size: 40px;
    font-family: Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif;
}

.repoList {
    margin-top: 80px;
    font-size: 40px;
    font-family: Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif;
}

.repoInfo {
    font-size: 20px;
    font-family: Impact, Haettenschweiler, 'Arial Narrow Bold', sans-serif;
    text-align: left;
    margin-left: 50px;
}

.githubLink {
    font-size: 20px;
    position: relative;
    color: rgb(5, 255, 38);
    top: 20%;
}