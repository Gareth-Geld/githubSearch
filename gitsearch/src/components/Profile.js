import '../css/Profile.css';
import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import 'bootstrap/dist/css/bootstrap.min.css';
import loading from '../images/loading.gif';
//This is the profile component This displays the users profile if their result was clicked
function Profile() {
    const { username } = useParams();
    const [error, setError] = useState('');
    const [backendData, setBackendData] = useState(null);
    const [repositories, setRepositories] = useState([]);

    const [isLoading, setIsLoading] = useState(true);
    const [loadingImg] = useState(loading);

    useEffect(() => {
        const fetchData = async () => {
            try {
                const response = await fetch(`/api/github/${username}`);

                if (response.ok) {
                    const data = await response.json();
                    console.log('Fetched data:', data); // Log the fetched data
                    setBackendData(data);

                    // Fetch user's repositories
                    const reposResponse = await fetch(data.repos_url);
                    if (reposResponse.ok) {
                        const reposData = await reposResponse.json();
                        console.log('Fetched repositories:', reposData);

                        // Get the first two repositories
                        const firstTwoRepos = reposData.slice(0, 2);
                        // Fetch detailed repository information for each repository
                        const detailedRepos = await Promise.all(
                            firstTwoRepos.map(async (repo) => {
                                const detailedRepoResponse = await fetch(repo.url);
                                if (detailedRepoResponse.ok) {
                                    const detailedRepoData = await detailedRepoResponse.json();

                                    // Fetch the last 5 commits for each repository
                                    const commitsResponse = await fetch(`${repo.url}/commits?per_page=5`);
                                    if (commitsResponse.ok) {
                                        const commitsData = await commitsResponse.json();
                                        detailedRepoData.commits = commitsData;
                                    }

                                    return detailedRepoData;
                                }
                                return null;
                            })
                        );

                        setRepositories(detailedRepos);
                        setIsLoading(false);
                    }
                } else {
                    console.error('Fetch failed with status:', response.status);
                    setError('User not found');
                    setIsLoading(false);
                }
            } catch (error) {
                console.error('An error occurred:', error);
                setError('An error occurred');
                setIsLoading(false);
            }
        };

        fetchData(); // Call the async function to fetch data
    }, [username]); // Include username as a dependency to re-run the effect when it changes
    //This renders the  Profile it checks if the backedData is there before rendering anything it just displays a loading if if the data is not ready to be presented
    return (
        <div className="profileContainer">
            {isLoading ? (
                <div className="loadingContainer">
                    <img src={loadingImg} alt="loading" srcset="" />
                </div>
            ) : backendData ? (
                <div>
                    <div className="row">
                        <div className="col-sm">
                            <img src={backendData.avatar_url} className='avatarProfile' alt="" srcSet="" />
                            <a href={backendData.html_url} target="_blank" rel="noopener noreferrer"><p className="githubLink">Github Profile</p></a>
                        </div>
                        <div className="col-sm"><p className='title'>{backendData.name || username}</p></div>
                    </div>
                    <div className="row">
                        <div className="col-sm">
                            <div className="userInfo">
                                <p className='BioText'>{backendData.bio || 'No bio available.'}</p>
                                <p>Followers: {backendData.followers || 0}</p>
                                <p>Following: {backendData.following || 0}</p>
                            </div>
                        </div>
                        <div className="col-sm">
                            <div className="repoList">
                                <h2 className='repoListTitle'>Repositories:</h2>
                                <ul>
                                    {repositories.map((repo) => (
                                        <li key={repo.id}>
                                            <a href={repo.html_url} target="_blank" rel="noopener noreferrer">
                                                {repo.name}
                                            </a>
                                        </li>
                                    ))}
                                </ul>
                            </div>
                        </div>
                    </div>
                    <div className="row">
                        <p className='repoTitle'>REPO'S</p>
                        {repositories.map((repo) => (
                            <div className="col-sm">
                                <div className="repoInfo">
                                    <div key={repo.id}>
                                        <h3>
                                            <a href={repo.html_url} target="_blank" rel="noopener noreferrer">
                                                {repo.name}
                                            </a>
                                        </h3>
                                        <p>Created: {new Date(repo.created_at).toLocaleDateString()}</p>
                                        <p>Last Commit: {new Date(repo.updated_at).toLocaleDateString()}</p>
                                        <p>Description: {repo.description || 'No description available.'}</p>

                                        <h4>Last 5 Commits:</h4>
                                        <ul>
                                            {repo.commits.map((commit) => (
                                                <li key={commit.sha}>{commit.commit.message}</li>
                                            ))}
                                        </ul>
                                    </div>
                                </div>
                            </div>
                        ))}

                    </div>

                </div>
            ) : (
                <p>{error}</p>
            )
            }
        </div >
    );
}

export default Profile;