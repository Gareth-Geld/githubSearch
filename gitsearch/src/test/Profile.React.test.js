import React from 'react';
import Profile from '../components/Profile';
import renderer from 'react-test-renderer';

test('renders correctly', () => {
    const tree = renderer
        .create(<Profile />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});