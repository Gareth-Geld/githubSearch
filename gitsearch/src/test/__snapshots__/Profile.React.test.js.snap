// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="profileContainer"
>
  <div
    className="loadingContainer"
  >
    <img
      alt="loading"
      src="loading.gif"
      srcset=""
    />
  </div>
</div>
`;
